EIPS/eip-5656.md

```diff
@@ -0,0 +1,93 @@
+masked, or'd, and stored again. This overhead is significant. One edge case is if the last "partial word" is a single byte,
+it can be efficiently stored using `MSTORE8`.
+
+As example use case, copying 256 bytes costs:
+- at least 757 gas pre-EIP-2929 using the identity precompile
+- at least 157 gas post-EIP-2929 using the identity precompile
+- at least 96 gas using unrolled `MLOAD`/`MSTORE` instructions
+- 27 gas using this EIP
+
+According to an analysis of blocks 10537502 to 10538702, roughly 10.5% of memory copies would have had improved performance with the
+availability of an `MCOPY` instruction.
+
+Memory copying is used by languages like Solidity and Vyper, where we expect this improvement to provide efficient means of building
+data structures, including efficient sliced access and copies of memory objects. Having a dedicated `MCOPY` instruction would also add
+forward protection against future gas cost changes to `CALL` instructions in general.
+
+Having a special `MCOPY` instruction makes the job of static analyzers and optimizers easier, since the effects of a `CALL` in general
+have to be fenced, whereas an `MCOPY` instruction would be known to only have memory effects. Even if special cases are added
+for precompiles, a future hard fork could change `CALL` effects, and so any analysis of code using the identity precompile would only
+be valid for a certain range of blocks.
+
+Finally, we expect memory copying to be immensely useful for various computationally heavy operations, such as EVM384,
+where it is identified as a significant overhead.
+
+## Specification
+
+Starting `BLOCK_NUMBER >= HF`, a new instruction, `MCOPY` (`0x5c`) is introduced.
+
+It takes three arguments off the stack (1. represents the top of the stack):
+1. `dst`
+2. `src`
+3. `length`
+
+It copies `length` bytes from the offset pointed at `src` to the offset pointed at `dst` in memory.
+Copying takes place as if an intermediate buffer were used, allowing the destination and source to overlap.
+
+If `length > 0` and (`src + length` or `dst + length`) is beyond the current memory length, the memory is extended with respective gas cost applied.
+
+The gas cost of this instruction mirrors that of other `Wcopy` instructions and is `Gverylow + Gcopy * ceil(length / 32)`.
+
+## Rationale
+
+Production implementation of exact-word memory copying and partial-word memory copying can be found in the Solidity, Vyper and Fe compilers.
+
+With [EIP-2929](https://eips.ethereum.org/EIPS/eip-2929) the call overhead using the identity precompile was reduced from 700 to 100 gas.
+This is still prohibitive for making the precompile a reasonable alternative again.
+
+## Backwards Compatibility
+
+This EIP introduces a new instruction which did not exists previously. Already deployed contracts using this instruction could change their behaviour after this EIP.
+
+## Test Cases
+
+TBA
```

Comment:
```suggestion
```

Reply:
If possible would like to keep as it is nice to include some core test cases and this is a reminder for us.

Reply:
I would be okay with this:
```suggestion
<!-- TODO: Test Cases -->
```

Reply:
Comments are not rendered, so they are not very useful.

Reply:
That's the point. They aren't rendered, so they aren't shown unless you're editing the source file.